helper/resource/state.go

```diff
@@ -82,6 +85,12 @@ func (conf *StateChangeConf) WaitForState() (interface{}, error) {
 		wait := 100 * time.Millisecond
 
 		for {
+			// Intentionally not deferring the Unlock
+			// if this function returns, I don't want to let the timeout
+			// code run; it should pick up on the channel closing
+			// this does mean, however, that any continue statements MUST call
+			// refreshMutex.Unlock() before continue
```

Comment:
Agree that it leaves the function easy to introducing new bugs, but this seemed like the one that was least likely to introduce race conditions (as I'm admittedly not terribly familiar with go concurrency and synchronization primitives, so I'm open to better ways of doing this). Basically, the two guarantees I'm attempting to write are:

1. Once the timeout has expired, `conf.Refresh` **cannot** be called again
1. The timeout error **cannot** be returned while `conf.Refresh` is still running, unless the grace period has expired.

This is also tricky because the returning inside the for loop is what triggers the closing of `doneCh` which is what allows the return value to be processed. Now I want to have some more complex behavior inside the for loop -- in some code paths, I want to close `doneCh` and in others I want to unlock `refreshMutex`. What I'm not sure of is the guarantees of go around if I close `doneCh` and unlock `refreshMutex` then what order will the `select` statements get processed?

Anyway, because I want either `conf.Refresh()` to be called or the timeout code to process, it seems like a mutex is the right primitive.